### PR TITLE
Add support for "graded" complexity reports

### PR DESCRIPTION
**Pull request by tiyberius**
Some tools such as [Radon](https://radon.readthedocs.org/en/latest/) output complexity reports with "graded" ratings A through F. It would be cool if the CCM plugin supported these tools out of the box. Even better, not captured in this commit because I have limited experience writing Jenkins plugins, would be to configure the CCM plugin to determine what complexity ranges map to "high, normal and low" priorities.

### CODE REVIEW

**Comment by tiyberius** [comment]
Please note that as of this time, I was not able to build the plugin because of https://issues.jenkins-ci.org/browse/JENKINS-10818. I'm getting a "Caused by: java.lang.NoClassDefFoundError: com/sun/mirror/apt/AnnotationProcessorFactory" and I can't figure out why. Please see https://jenkins.ci.cloudbees.com/job/plugins/job/ccm-plugin/4/ for full details.

**Comment by tiyberius** [comment]
The more I think about it, the CCM plugin should ship with some sensible defaults, but allow the user to determine which complexities map to what "priority". On the subject of priorities, not sure if this is the CCM Plugin or the Static Code Analysis Plugin, but when displaying code complexities, wouldn't it make sense to swap the term "priority" for "complexity"? So you'd have "high, normal, and low" complexities instead of priorities? Because as far as I'm concerned any files that are of "low complexity" are really of "NO priority", not "low priority". Anyway, just wondering if anyone shares those sentiments.

**Comment by tiyberius** [comment]
Okay, I successfully built the plugin using Java 6. Looks like the cloud CI server is using Java 7 ... not sure what to change in the pom.xml or other files to make it compatible with Java7

**Comment by kinow** [comment]
Hi @tiyberius, working on other OSS projects at the moment, but I'll take a look at your PR tonight :-) Feel free to update it in the meantime if necessary. I'll probably need a few hours to refresh my memory on the CCM plug-in codebase, though.

**Comment by tiyberius** [comment]
Thanks for the response! As I mentioned in one of my comments above, this PR might not be a great solution, a semi-hack if you will. I think ideally you could set some sensible defaults for what complexities map to what priorities, but ultimately let the user change their threshold about high/normal/low priorities. Don't know how much work is involved there.

**Comment by kinow** [comment]
Thanks for taking your time to report the issue and send a PR. If it requires writing some Java code, or Jelly UI for that, I can probably help. But choosing - if that's really the case - the default values sounds tricky to me. Would you have some suggestion for that?

**Comment by tiyberius** [comment]
Based on personal experience, a CCM of 1-5 is excellent, no refactoring should be necessary ("low" priority if that terminology stays). 6-10 is okay, starting to get a little complex ("normal" priority if you will). I think anything beyond 10 should be a "high" priority/complexity. Those rangers are based on my personal experience. It may seem a little restrictive for general use, so you could loosen them up a bit. The CC Score, Rank and Risk table on this page https://radon.readthedocs.org/en/latest/commandline.html might be of interest

**Comment by kinow** [comment]
Excellent, we already have something to start. Thanks @tiyberius

**Comment by tiyberius** [comment]
No problem! Please let me know if I can offer any additional help

**Comment by kinow** [comment]
>Even better, not captured in this commit because I have limited experience writing Jenkins plugins, would be to configure the CCM plugin to determine what complexity ranges map to "high, normal and low" priorities.

This would probably be better at another plug-in. Last time I worked on CCM, IIRC, I added support to the [static code analysis plugin](https://wiki.jenkins-ci.org/display/JENKINS/Static+Code+Analysis+Plug-ins). It would probably be a good idea ping the maintainer (via Jenkins JIRA) and see if s/he thinks it is a good idea too.

**Comment by kinow** [comment]
> On the subject of priorities, not sure if this is the CCM Plugin or the Static Code Analysis Plugin,

Sorry, hadn't seen your comment :-) sorry.

File: src/main/java/hudson/plugins/ccm/parser/CcmParser.java [hunk 1]
```diff
@@ -141,4 +134,43 @@ public CcmParser(String defaultEncoding) {
...
+
+		return priority;
+	}
+
+	private boolean isMetricHighPriority(Metric metric){
+		String metricClassification = metric.getClassification();
+		if (metricClassification.contains("high")){
+			return true;
+		}
+		else if (metricClassification.contentEquals("C") || metricClassification.contentEquals("D") ||
```
Comment (kinow) [inline src/main/java/hudson/plugins/ccm/parser/CcmParser.java]: Hi, do you have an updated CCM output that emits these C, D, E, F value? The only one that I have (stopped using/building for windows long time ago) is this one - https://github.com/jenkinsci/ccm-plugin/blob/master/src/test/resources/hudson/plugins/ccm/parser/ccm.result.xml

Comment (tiyberius) [inline src/main/java/hudson/plugins/ccm/parser/CcmParser.java]: Sure! I ran the Radon complexity tool against a Python module called PynamoDB and updated the PR. You can see the resulting output here https://github.com/jenkinsci/ccm-plugin/blob/a5d795186e747557c54dcd131739bbfc308aa11e/src/test/resources/hudson/plugins/ccm/parser/pynamodb_ccm.xml

Comment (kinow) [inline src/main/java/hudson/plugins/ccm/parser/CcmParser.java]: :-O

> This is specifically targeted to Jenkin’s plugin CCM.

I had no idea Radon was outputting CCM XML :-) I'll include your XML as a test case later today. 

Thanks a lot!

**Comment by kinow** [comment]
Merged! Thank you for your contribution!

Would you like to get your name added to pom.xml? If so just send me your details and I'll update it. Can't promise, but will try to write a test case and then cut a new release tonight, otherwise will do so tomorrow. 

Hopefully by next week you'll be able to update your plug-in and use the graded complexity report :-)

**Comment by tiyberius** [comment]
Ah yes, the tests. I was going to write test cases initially but wasn't sure if the code as gonna go through. I can update the pom.xml with my details, I think that'll be easier. I won't get around to it until Sunday, but I can write the tests if you want!

**Comment by kinow** [comment]
It'd be great if you can do that! 
That way it will be definitely easier to release it before Monday.
Thank you

**Pull request merged**